src/pe/mod.rs

```diff
@@ -267,17 +277,180 @@ impl<'a> PE<'a> {
+                        ))?;
+                bytes.gwrite_with(coff_strings.len() as u32, &mut offset, ctx)?;
+                bytes.gwrite(coff_strings.bytes, &mut offset)?;
+                max_offset = max(offset, max_offset);
+            }
+        }
+        if let Some(opt_header) = self.header.optional_header {
+            // Takes care of:
+            // - export table (.edata)
+            // - import table (.idata)
+            // - bound import table
+            // - import address table
+            // - delay import tables
+            // - resource table (.rsrc)
+            // - exception table (.pdata)
+            // - base relocation table (.reloc)
+            // - debug table (.debug)
+            // - load config table
+            // - tls table (.tls)
+            // - architecture (reserved, 0 for now)
+            // - global ptr is a "empty" data directory (header-only)
+            // - clr runtime header (.cormeta is object-only)
+            for (dt_type, dd) in opt_header.data_directories.dirs() {
+                // - attribute certificate table is a bit special
+                // its size is not the real size of all certificates
+                // you can check the parse logic to understand how that works
+                if dt_type == DataDirectoryType::CertificateTable {
+                    let mut certificate_start = dd.virtual_address.try_into()?;
+                    for certificate in &self.certificates {
+                        bytes.gwrite_with(certificate, &mut certificate_start, ctx)?;
+                        max_offset = max(max_offset, certificate_start);
+                    }
+                } else {
+                    let offset = dd.offset.ok_or(Self::Error::Malformed(
+                        "Data directory was not read with offset information, cannot rewrite"
+                            .into(),
+                    ))?;
+                    let dd_written = bytes.pwrite(dd.data(&self.bytes)?, offset)?;
+                    max_offset = max(max_offset, offset);
+                    debug!(
+                        "writing {:?} at {} for {} bytes",
+                        dt_type, offset, dd_written
+                    );
+                }
+            }
+        }
+
+        Ok(max_offset)
+    }
+}
+
 /// An analyzed COFF object
 #[derive(Debug)]
 pub struct Coff<'a> {
     /// The COFF header
     pub header: header::CoffHeader,
     /// A list of the sections in this COFF binary
     pub sections: Vec<section_table::SectionTable>,
-    /// The COFF symbol table.
-    pub symbols: symbol::SymbolTable<'a>,
-    /// The string table.
-    pub strings: strtab::Strtab<'a>,
+    /// The COFF symbol table, they are not guaranteed to exist.
+    /// For an image, this is expected to be None as COFF debugging information
+    /// has been deprecated.
+    pub symbols: Option<symbol::SymbolTable<'a>>,
```

Comment:
breaking changes too

Reply:
Yep, I think those are necessary though.

Reply:
yea mostly commenting for when i write changelog :)